pkg/apply/apply.go

```diff
@@ -68,15 +68,21 @@ func ApplyObject(ctx context.Context, cli k8sclient.Client, objState objectstate
 		return objState.Desired, true, nil
 	}
 
+	// set default to the desired object to avoid the comparison from failing
+	// on default values which are set by the API server
+	desiredWithDefault := objState.Default(objState.Existing, objState.Desired)
+
 	// Merge the desired object with what actually exists
-	merged, err := objState.Merge(objState.Existing, objState.Desired)
+	merged, err := objState.Merge(objState.Existing, desiredWithDefault)
```

Comment:
another option could be to make our merge functions smarter. We should only copy the desired fields in the existing object. This allows us to avoid defaults (and the need to depend on defaulter) and was the original intent of the `merge` package, which was however lost in the initial implementation.

We can take the chance to rethink the `Compare` step. Can we do `Merge+Compare` in one single go? Can the merging step infer if a field we care about changed during the merge?

Reply:
Defining a list of fields we care about using a black/white list is possible however it has its own cons:
1. We need to define the fields we care about which is not that easy because we don't know what will happen in case that a field "we didn't care about" is misconfigured, could it break anything?
2. We need to write a custom compare/merge functions that will be smart enough to skip and focus on the fields according to our needs, which is basically similar to setting the defaults and goes through everything.
3. What should we do if user is tampering with fields we didn't care about and messing things up? we'll never be able to fix it automatically, a manual intervention will be needed.